src/runtime-rs/crates/resource/src/cdi_devices/container_device.rs

```diff
@@ -0,0 +1,96 @@
+use std::path::Path;
+
+use anyhow::Result;
+use oci_spec::runtime::Spec;
+
+use super::{ContainerDevice, CDI_DEVICE_VENDOR_CLASS};
+use agent::types::Device;
+
+const CDI_PREFIX: &str = "cdi.k8s.io";
+
+fn get_vendor_class<'a>(vendor_id: &'a str, class_id: &'a str) -> Option<&'a str> {
+    let vendor_class = format!("{}-{}", vendor_id, class_id);
+    // The first 12 characters of the string ("0x10de-0x030") provide a concise
+    // and clear identification of both the manufacturer and the device category.
+    CDI_DEVICE_VENDOR_CLASS.get(&vendor_class[..12]).copied()
+}
+
+// Extract guest_pcipath from device_options
+fn extract_guest_pcipath(options: &[String]) -> Option<String> {
+    options
+        .first()
+        .and_then(|option| option.split('=').nth(1))
+        .map(|path| path.to_string())
+}
+
+// Sort devices by guest_pcipath
+fn sort_devices_by_guest_pcipath(devices: &mut [ContainerDevice]) {
+    devices.sort_by(|a, b| {
+        let guest_path_a = extract_guest_pcipath(&a.device.options);
+        let guest_path_b = extract_guest_pcipath(&b.device.options);
+
+        guest_path_a.cmp(&guest_path_b)
+    });
+}
+
+// Annotate container device
+pub fn annotate_container_device(
+    spec: &mut Spec,
+    container_devices: Vec<ContainerDevice>,
+) -> Result<Vec<Device>> {
+    let mut devices_agent: Vec<Device> = Vec::new();
+
+    // Step 1: Extract all devices and filter out devices without device_info for vfio_devices
+    let vfio_devices: Vec<ContainerDevice> = container_devices
+        .into_iter()
+        .map(|device| {
+            // push every device's Device to agent_devices
+            devices_agent.push(device.device.clone());
+            device
+        })
+        .filter(|device| device.device_info.is_some())
+        .collect();
+
+    // Step 2: Group devices by vendor_id-class_id
+    let mut grouped_devices: HashMap<String, Vec<ContainerDevice>> = HashMap::new();
+    for device in vfio_devices {
+        // Extract the vendor/class key and insert into the map if both are present
+        if let Some(key) = device
+            .device_info
+            .as_ref()
+            .and_then(|info| get_vendor_class(&info.vendor_id, &info.class_id))
+        {
+            grouped_devices
+                .entry(key.to_owned())
+                .or_default()
+                .push(device);
+        }
+    }
+
+    // Step 3: Sort devices within each group by guest_pcipath
+    grouped_devices
+        .iter_mut()
+        .for_each(|(vendor_class, devices)| {
+            sort_devices_by_guest_pcipath(devices);
+            devices.iter().enumerate().for_each(|(index, device)| {
+                if let Some(iommu_grpid) = Path::new(&device.device.container_path)
+                    .file_name()
+                    .and_then(|name| name.to_str())
+                {
+                    spec.annotations_mut().as_mut().unwrap().insert(
+                        format!("{}/vfio{}", CDI_PREFIX, iommu_grpid),
```

Comment:
This will only work for IOMMU groups with one device. If you e.g. have a GPU in graphics mode you would have a multi function device. If you pass-through `0000:19:00.00` GPU and `0000:19:00.01` Audio this will map to the same IOMMU group, but different devices. You need to add the device number to the `iommu_grpid` e.g. `cdi.k8s.io/vfio17.0`

Reply:
yeah, good catch. the vfio with multi-functions is needed for GPU cases